packages/std/src/exports.rs

```diff
@@ -44,154 +46,196 @@ extern "C" fn deallocate(pointer: u32) {
-pub fn do_query<T: DeserializeOwned + JsonSchema>(
-    query_fn: &dyn Fn(
-        &Extern<ExternalStorage, ExternalApi, ExternalQuerier>,
-        T,
-    ) -> StdResult<QueryResponse>,
-    msg_ptr: u32,
-) -> u32 {
-    let res: QueryResult = _do_query(query_fn, msg_ptr as *mut Region);
+    let res = _do_handle(handle_fn, env_ptr as *mut Region, msg_ptr as *mut Region);
     let v = to_vec(&res).unwrap();
     release_buffer(v) as u32
 }
 
 /// do_migrate should be wrapped in an external "C" export, containing a contract-specific function as arg
-pub fn do_migrate<T, U>(
+///
+/// - `M`: message type for request
+/// - `C`: custom response message type (see CosmosMsg)
+/// - `E`: error type for responses
+pub fn do_migrate<M, C, E>(
     migrate_fn: &dyn Fn(
         &mut Extern<ExternalStorage, ExternalApi, ExternalQuerier>,
         Env,
-        T,
-    ) -> MigrateResult<U>,
+        M,
+    ) -> Result<MigrateResponse<C>, E>,
     env_ptr: u32,
     msg_ptr: u32,
 ) -> u32
 where
-    T: DeserializeOwned + JsonSchema,
-    U: Serialize + Clone + fmt::Debug + PartialEq + JsonSchema,
+    M: DeserializeOwned + JsonSchema,
+    C: Serialize + Clone + fmt::Debug + PartialEq + JsonSchema,
+    E: ToString,
 {
-    let res: MigrateResult<U> =
-        _do_migrate(migrate_fn, env_ptr as *mut Region, msg_ptr as *mut Region);
+    let res = _do_migrate(migrate_fn, env_ptr as *mut Region, msg_ptr as *mut Region);
     let v = to_vec(&res).unwrap();
     release_buffer(v) as u32
 }
 
-fn _do_init<T, U>(
+/// do_query should be wrapped in an external "C" export, containing a contract-specific function as arg
+///
+/// - `M`: message type for request
+/// - `E`: error type for responses
+pub fn do_query<M, E>(
+    query_fn: &dyn Fn(
+        &Extern<ExternalStorage, ExternalApi, ExternalQuerier>,
+        M,
+    ) -> Result<QueryResponse, E>,
+    msg_ptr: u32,
+) -> u32
+where
+    M: DeserializeOwned + JsonSchema,
+    E: ToString,
+{
+    let res = _do_query(query_fn, msg_ptr as *mut Region);
+    let v = to_vec(&res).unwrap();
+    release_buffer(v) as u32
+}
+
+fn _do_init<M, C, E>(
     init_fn: &dyn Fn(
         &mut Extern<ExternalStorage, ExternalApi, ExternalQuerier>,
         Env,
-        T,
-    ) -> InitResult<U>,
+        M,
+    ) -> Result<InitResponse<C>, E>,
     env_ptr: *mut Region,
     msg_ptr: *mut Region,
-) -> InitResult<U>
+) -> ContractResult<InitResponse<C>>
 where
-    T: DeserializeOwned + JsonSchema,
-    U: Serialize + Clone + fmt::Debug + PartialEq + JsonSchema,
+    M: DeserializeOwned + JsonSchema,
+    C: Serialize + Clone + fmt::Debug + PartialEq + JsonSchema,
+    E: ToString,
 {
     let env: Vec<u8> = unsafe { consume_region(env_ptr) };
     let msg: Vec<u8> = unsafe { consume_region(msg_ptr) };
-    let env: Env = from_slice(&env)?;
-    let msg: T = from_slice(&msg)?;
+    let env: Env = try_into_contract_result!(from_slice(&env));
```

Comment:
Okay, not sure what alternative would work here. I cannot really argue against the macro

Reply:
The only alternative is unrolling this into a `mach` statement that returns on `Err` and contunues on `Ok`. We can do that, no problem. It would retult in 7 match more statements in this file.